File: backend/goat_listing/src/main/java/com/goat/cart/application/usecases/RemoveItemFromCartUseCase.java
package com.goat.cart.application.usecases;

import com.goat.cart.adapters.external.RestCartServiceAdapter;
import com.goat.cart.domain.exceptions.ItemNotFoundException;
import com.goat.cart.ports.CartRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.UUID;

/**
 * Use case para remover un item del carrito.
 * 
 * Usa el adaptador HTTP directamente para comunicarse con el servicio Python
 */
@Component
public class RemoveItemFromCartUseCase {

    private static final Logger logger = LoggerFactory.getLogger(RemoveItemFromCartUseCase.class);

    private final CartRepository cartRepository;
    private final RestCartServiceAdapter cartServiceAdapter;

    public RemoveItemFromCartUseCase(CartRepository cartRepository, RestCartServiceAdapter cartServiceAdapter) {
        this.cartRepository = cartRepository;
        this.cartServiceAdapter = cartServiceAdapter;
    }

    public void execute(UUID userId, UUID itemId) {
        logger.info("Removiendo item {} del carrito del usuario {}", itemId, userId);

        // 1. Verificar que el carrito existe y contiene el item
        var cart = cartRepository.findActiveByUser(userId)
                .orElseThrow(() -> new RuntimeException("Carrito no encontrado"));

        var itemExists = cart.getItems().stream()
                .anyMatch(i -> i.getId().equals(itemId));

        if (!itemExists) {
            throw new ItemNotFoundException("El item no se encuentra en el carrito");
        }

        // 2. Remover el item directamente en el servicio Python
        cartServiceAdapter.removeItem(userId, itemId);

        logger.info("Item {} removido del carrito del usuario {}", itemId, userId);
    }
}

File: backend/goat_catalog/src/goat_catalog/cart/application/use_cases/remove_item_from_cart_use_case.py
"""Caso de uso para remover un item del carrito."""

import logging
from uuid import UUID

from ...domain.exceptions import CartNotFoundException
from ...domain.repositories import CartRepository

logger = logging.getLogger(__name__)


class RemoveItemFromCartUseCase:
    """Caso de uso para remover un item del carrito."""

    def __init__(self, cart_repository: CartRepository) -> None:
        """Inicializa el caso de uso.

        Args:
            cart_repository: Repositorio para acceder a los carritos
        """
        self._cart_repository = cart_repository

    async def execute(self, user_id: UUID, item_id: UUID) -> None:
        """Remueve un item del carrito del usuario.

        Args:
            user_id: ID del usuario
            item_id: ID del item a remover

        Raises:
            CartNotFoundException: Si el carrito no existe
            ValueError: Si el item no está en el carrito
        """
        logger.info("Removiendo item %s del carrito del usuario %s", item_id, user_id)

        # 1. Obtener carrito del usuario
        cart = await self._cart_repository.find_active_by_user(user_id)
        if cart is None:
            logger.warning("Carrito no encontrado para usuario %s", user_id)
            raise CartNotFoundException("Carrito no encontrado")

        # 2. Verificar que el item existe
        item_exists = any(item.id == item_id for item in cart.items)
        if not item_exists:
            logger.warning("Item %s no encontrado en el carrito del usuario %s", item_id, user_id)
            raise ValueError(f"El item {item_id} no se encuentra en el carrito")

        # 3. Remover el item
        cart.remove_item(item_id)

        # 4. Guardar cambios
        await self._cart_repository.save(cart)

        logger.info("Item %s removido del carrito del usuario %s", item_id, user_id)

